oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/mongo/MongoDBExceptionTest.java

```diff
@@ -132,8 +136,132 @@ public void idInExceptionMessage() {
+            assertThat(e.getMessage(), containsString(exceptionMsg));
+            assertThat(e.getMessage(), containsString(docName));
+        }
+    }
+
+    @Test
+    public void multiCreateOrUpdate16MBDoc() {
+
+        List<UpdateOp> updateOps = new ArrayList<>();
+
+        UpdateOp op = new UpdateOp("/test", true);
+        op = create1MBProp(op);
+
+        store.createOrUpdate(Collection.NODES, op);
+
+        UpdateOp op1 = new UpdateOp("/foo", true);
+        op1 = create16MBProp(op1);
+
+        // Updating both doc with 16MB
+        op = create16MBProp(op);
+        updateOps.add(op);
+        updateOps.add(op1);
+        exceptionMsg = "Resulting document after update is larger than 16777216";
+
+        try {
+            store.createOrUpdate(Collection.NODES, updateOps);
+            fail("DocumentStoreException expected");
+        } catch (DocumentStoreException e) {
+            assertThat(e.getMessage(), containsString(exceptionMsg));
+        }
+    }
+
+    @Test
+    public void create16MBDoc() {
+
+        List<UpdateOp> updateOps = new ArrayList<>();
+
+        UpdateOp op1 = new UpdateOp("/test", true);
+        op1 = create1MBProp(op1);
+
+        UpdateOp op2 = new UpdateOp("/foo", false);
+        op2 = create1MBProp(op2);
+        op2 = create16MBProp(op2);
+
+        updateOps.add(op1);
+        updateOps.add(op2);
+        String id = op2.getId();
+        exceptionMsg = "Payload document size is larger than maximum of 16777216.";
+        try {
+            store.create(Collection.NODES, updateOps);
+            fail("DocumentStoreException expected");
+        } catch (DocumentStoreException e) {
+            assertThat(e.getMessage(), containsString(exceptionMsg));
+        }
+    }
+
+    @Test
+    public void findAndUpdate16MBDoc() throws Exception {
+        UpdateOp op = new UpdateOp("/foo", true);
+        op = create1MBProp(op);
+        store.createOrUpdate(Collection.NODES, op);
+        op = create16MBProp(op);
+        exceptionMsg = "Resulting document after update is larger than 16777216";
+        try {
+            store.findAndUpdate(Collection.NODES, op);
+            fail("DocumentStoreException expected");
+        } catch (DocumentStoreException e) {
+            assertThat(e.getMessage(), containsString(exceptionMsg));
+       }
+    }
+
     private void setExceptionMsg() {
         client.setExceptionBeforeUpdate(exceptionMsg);
         client.setExceptionBeforeQuery(exceptionMsg);
     }
+
+    private UpdateOp create1MBProp(UpdateOp op) {
+        // create a 1 MB property
+        String content = create1MBContent();
+        op.set("property0", content);
+        return op;
+    }
+
+    private UpdateOp create16MBProp(UpdateOp op) {
+        // create a 1 MB property
+        String content = create1MBContent();
+        op.set("property0", content);
+
+        //create 16MB property
+        for (int i = 1; i < 16; i++) {
+            op.set("property"+ i, content);
+        }
```

Comment:
```suggestion
        

        //create 16MB property
        for (int i = 0; i < 16; i++) {
            op.set("property"+ i, content);
        }
```

Reply:
IIUC then this one is still open

Reply:
somehow, one of my commits had removed it, added